src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -15,26 +15,122 @@ import scala.annotation.tailrec
 
   val blockRewardCalculator = new BlockRewardCalculator(blockchainConfig.monetaryPolicyConfig)
 
-  def executeBlock(
-    block: Block,
-    validators: Validators): Either[BlockExecutionError, Seq[Receipt]] = {
+  def importBlock(block: Block): BlockImportResult = {
+    // TODO: ignore blocks that are too far behind the head
+    if (blockchain.getBlockByHash(block.header.hash).isDefined)
+      ImportFailedBlockExists
+
+    else
+      blockchain.getBlockByHash(block.header.parentHash) match {
+        case None =>
+          ImportFailedMissingParent
+
+        case Some(parent) =>
+          executeBlock(block) match {
+            case Right(receipts) =>
+              val newTd = blockchain.getTotalDifficultyByHash(parent.header.hash).getOrElse(BigInt(0)) +
+                parent.header.difficulty
+
+              blockchain.save(block)
+              blockchain.save(block.header.hash, receipts)
+              blockchain.save(block.header.hash, newTd)
+
+              val isTop = isTopOfChain(block, newTd)
+              if (isTop) {
+                saveBranchNumbers(block.header)
+                blockchain.saveBestBlockNumber(block.header.number)
+              }
+
+              BlockImported(receipts, newTd, isTop)
+
+            case Left(error) =>
+              ImportFailedExecError(error)
+          }
+      }
+  }
+
+  private def isTopOfChain(block: Block, td: BigInt): Boolean = {
+    val bestBlock = blockchain.getBlockByNumber(blockchain.getBestBlockNumber()).get // TODO: safe?
+    val isChild = bestBlock.header.hash == block.header.parentHash
+    lazy val isHigherTd = td > blockchain.getTotalDifficultyByHash(bestBlock.header.hash).getOrElse(0)
+
+    isChild || isHigherTd
+  }
+
+  private def saveBranchNumbers(topHeader: BlockHeader): Unit = {
+    blockchain.saveBlockNumber(topHeader.number, topHeader.hash)
+
+    // we only save bock numbers for blocks at the top of the chain so both parents are guaranteed to exist
+    val parentByHash = blockchain.getBlockHeaderByHash(topHeader.parentHash).get
+    val parentByNumber = blockchain.getBlockHeaderByNumber(topHeader.number - 1).get
+
+    if (parentByNumber.hash != parentByHash.hash) {
+      saveBranchNumbers(parentByHash)
+    }
+  }
+
+  def resolveBranch(headers: Seq[BlockHeader]): BranchResolutionResult = {
+    if (!checkHeadersChain(headers))
+      InvalidBranch
+    else {
+      val parentByNumber = blockchain.getBlockHeaderByNumber(headers.head.number - 1)
+
+      parentByNumber match {
+        case None =>
+          UnknownBranch
+
+        case Some(parent) =>
+          // find blocks with same numbers in the current chain
+          val oldBranch = getOldBlocks(headers)
+          val currentBranchTotalDifficulty = oldBranch.map(_.header.difficulty).sum
+          val newBranchTotalDifficulty = headers.map(_.difficulty).sum
+
+          if (currentBranchTotalDifficulty < newBranchTotalDifficulty)
+            NewBranch(oldBranch)
```

Comment:
it seems a bit unintuitive, I mean the naming. How about something like: BranchReplaced or ChainReorganized or similar.
Same for `OldBranch` maybe rename to `ChainUntouched` or something like that